### ISSUE TITLE
quadratic or worse behavior in indenting/coloring

### ISSUE THREAD

**Issue by shriram** [brownplt/pyret-lang#630]
```
elementsFormal =
  [list: "H", "He", "Li", "Be", "B", "C", "N", "O", "F", "Ne", "Na", "Mg", "Al", "Si", "P", "S", "Cl", "Ar", "K", "Ca", "Sc", "Ti", "V", "Cr", "Mn", "Fe", "Co", "Ni", "Cu", "Zn", "Ga", "Ge", "As", "Se", "Br", "Kr", "Rb", "Sr", "Y", "Zr", "Nb", "Mo", "Tc", "Ru", "Rh", "Pd", "Ag", "Cd", "In", "Sn", "Sb", "Te", "I", "Xe", "Cs", "Ba", "La", "Ce", "Pr", "Nd", "Pm", "Sm", "Eu", "Gd", "Tb", "Dy", "Ho", "Er", "Tm", "Yb", "Lu", "Hf", "Ta", "W", "Re", "Os", "Ir", "Pt", "Au", "Hg", "Tl", "Pb", "Bi", "Po", "At", "Rn", "Fr", "Ra", "Ac", "Th", "Pa", "U", "Np", "Pu", "Am", "Cm", "Bk", "Cf", "Es", "Fm", "Md", "No", "Lr", "Rf", "Db", "Sg", "Bh", "Hs", "Mt", "Ds", "Rg", "Cn", "Uut", "Fl", "Uup", "Lv", "Uus", "Uuo"]
```
(Treat this as a single line, without line breaks in the middle of the list.)

Anywhere I try to just navigate in this line of code takes several seconds. There is almost certainly some non-linear behavior here; there's no good reason for it to take so long.

**Comment by blerner** [comment]
Yikes, that's bad indeed.  I think the problem seems to be related to @belph 's new bracket-matching code, which requires reading tokens more frequently than CM is optimized to do.  We might be missing an opportunity for caching the tokens we read, or something, but there's definitely a problem here.

**Comment by belph** [comment]
I'll take a look at this and post again once I have a better idea of where the inefficient behavior is coming from.

**Comment by belph** [comment]
Update: The problem is almost certainly that [this snippet](https://github.com/belph/CodeMirror/blob/8b09097b9396c08f724bdf9d8aaa7ae8dbdfbbf5/lib/codemirror.js#L4093-L4097) of code runs on each call to `CodeMirror.prototype.getTokenAt`, which causes the whole line to be re-tokenized. I am in the process of cleaning up the Pyret files (mainly the ones I've made) prior to migrating to the latest version of CodeMirror, which will enable us to use `CodeMirror.prototype.getLineTokens` and cache that information. This will hopefully fix the quadratic behavior.

I'll post with an update once I've got that working.
